drivers/gpio/gpio_silabs_siwx917.c

```diff
@@ -0,0 +1,418 @@
+{
+	const struct gpio_siwx917_port_config *cfg = port->config;
+
+	*value = sl_gpio_get_port_input(cfg->hal_port);
+
+	return 0;
+}
+
+static int gpio_siwx917_port_set_masked(const struct device *port, gpio_port_pins_t mask,
+					gpio_port_value_t value)
+{
+	const struct gpio_siwx917_port_config *cfg = port->config;
+	const struct device *parent = cfg->parent;
+	const struct gpio_siwx917_common_config *pcfg = parent->config;
+
+	/* Cannot use HAL function sl_gpio_set_port_output_value(), as it doesn't clear bits. */
+	pcfg->reg->PORT_CONFIG[cfg->port].PORT_LOAD_REG =
+		(pcfg->reg->PORT_CONFIG[cfg->port].PORT_LOAD_REG & ~mask) | (value & mask);
+
+	return 0;
+}
+
+static int gpio_siwx917_port_set_bits(const struct device *port, gpio_port_pins_t pins)
+{
+	const struct gpio_siwx917_port_config *cfg = port->config;
+
+	sl_gpio_set_port_output(cfg->hal_port, pins);
+
+	return 0;
+}
+
+static int gpio_siwx917_port_clear_bits(const struct device *port, gpio_port_pins_t pins)
+{
+	const struct gpio_siwx917_port_config *cfg = port->config;
+
+	sl_gpio_clear_port_output(cfg->hal_port, pins);
+
+	return 0;
+}
+
+static int gpio_siwx917_port_toggle_bits(const struct device *port, gpio_port_pins_t pins)
+{
+	const struct gpio_siwx917_port_config *cfg = port->config;
+
+	sl_gpio_toggle_port_output(cfg->hal_port, pins);
+
+	return 0;
+}
+
+static bool receiver_enabled(bool ulp, sl_gpio_port_t port, int pin)
+{
+	if (ulp) {
+		return ULP_PAD_CONFIG_REG & BIT(pin);
+	} else {
+		return PAD_REG(((port << 4) | pin))->GPIO_PAD_CONFIG_REG_b.PADCONFIG_REN;
+	}
+}
+
+int gpio_siwx917_port_get_direction(const struct device *port, gpio_port_pins_t map,
+				    gpio_port_pins_t *inputs, gpio_port_pins_t *outputs)
+{
+	const struct gpio_siwx917_port_config *cfg = port->config;
+
+	if (inputs != NULL) {
+		*inputs = 0;
+	}
+	if (outputs != NULL) {
+		*outputs = 0;
+	}
+	for (int i = 0; i < 16; i++) {
+		if ((map & BIT(i))) {
+			if (sl_si91x_gpio_get_pin_direction(cfg->hal_port, i) == 0) {
+				if (outputs != NULL) {
+					*outputs |= BIT(i);
+				}
+			}
+			if (receiver_enabled(cfg->ulp, cfg->port, i)) {
+				if (inputs != NULL) {
+					*inputs |= BIT(i);
+				}
+			}
+		}
+	}
+	return 0;
+}
+
+static int gpio_siwx917_manage_callback(const struct device *port, struct gpio_callback *callback,
+					bool set)
+{
+	struct gpio_siwx917_port_data *data = port->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static int gpio_siwx917_interrupt_configure(const struct device *port, gpio_pin_t pin,
+					    enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	const struct gpio_siwx917_port_config *cfg = port->config;
+	const struct device *parent = cfg->parent;
+	const struct gpio_siwx917_common_config *pcfg = parent->config;
+	struct gpio_siwx917_common_data *data = parent->data;
+	sl_si91x_gpio_interrupt_config_flag_t flags = 0;
+
+	if (mode & GPIO_INT_DISABLE) {
+		for (size_t i = 0; i < INTERRUPT_COUNT; i++) {
```

Comment:
I suggest `ARRAY_SIZE(data->interrupts)` rather than `INTERRUPT_COUNT`.

Reply:
Done. Ended up using ARRAY_FOR_EACH() for the entire loop definition.